vsintegration/src/FSharp.Editor/Common/Extensions.fs

```diff
@@ -40,6 +42,75 @@ type Document with
                 languageServices.GetService<'T>()
                 |> Some
 
+module private SourceText =
+
+    open System.Runtime.CompilerServices
+
+    let weakTable = ConditionalWeakTable<SourceText, ISourceText>()
+
+    let create (sourceText: SourceText) =
+
+        let sourceText =
+            { new ISourceText with
+            
+                member __.Item with get index = sourceText.[index]
+
+                member __.GetLineString(lineIndex) =
+                    sourceText.Lines.[lineIndex].ToString()
+
+                member __.GetLineCount() =
+                    sourceText.Lines.Count
+
+                member __.GetLastCharacterPosition() =
+                    if sourceText.Lines.Count > 0 then
+                        (sourceText.Lines.Count, sourceText.Lines.[sourceText.Lines.Count - 1].Span.Length)
+                    else
+                        (0, 0)
+
+                member __.GetSubTextString(start, length) =
+                    sourceText.GetSubText(TextSpan(start, length)).ToString()
+
+                member __.SubTextEquals(target, startIndex) =
+                    if startIndex < 0 || startIndex >= sourceText.Length then
+                        raise (ArgumentOutOfRangeException("startIndex"))
+
+                    if String.IsNullOrEmpty(target) then
+                        raise (ArgumentException("Target is null or empty.", "target"))
+
+                    let lastIndex = startIndex + target.Length
+                    if lastIndex <= startIndex || lastIndex >= sourceText.Length then
+                        raise (ArgumentException("Target is too big.", "target"))
```

Comment:
Nit: prefer invalidArg as per [guidelines](https://docs.microsoft.com/en-us/dotnet/fsharp/style-guide/conventions#use-exceptions-when-errors-cannot-be-represented-with-types)